### PR TITLE
UCT/IB: Enhanced CQE zipping support

### PR DESCRIPTION
**Pull request by Artemy-Mellanox**

### CODE REVIEW

File: src/uct/ib/rc/accel/rc_mlx5.inl [hunk 1]
```diff
@@ -267,21 +267,31 @@ uct_rc_mlx5_iface_poll_rx_cq(uct_rc_mlx5_iface_common_t *iface, int poll_flags)
...
 
-    idx    = cq->cq_ci;
-    cqe    = uct_ib_mlx5_get_cqe(cq, idx);
-    op_own = cqe->op_own;
+    idx = cq->cq_ci;
+    cqe = uct_ib_mlx5_get_cqe(cq, idx);
 
-    if (ucs_unlikely(uct_ib_mlx5_cqe_is_hw_owned(op_own, idx, cq->cq_length))) {
+    if ((poll_flags & UCT_IB_MLX5_POLL_FLAG_CQE_ZIP) && cq->zip) {
+        is_hw_owned = (cqe->signature != ((idx / cq->cq_length) % 256));
```
Comment (Artemy-Mellanox) [inline src/uct/ib/rc/accel/rc_mlx5.inl:279]: (u8)cqe->signature != (u8)(idx >> cq->log_cq_length)

File: src/uct/ib/rc/accel/rc_mlx5_iface.c [hunk 2]
```diff
@@ -175,24 +176,32 @@ uct_rc_mlx5_iface_progress(void *arg, int flags)
...
+    return count + uct_rc_mlx5_iface_poll_tx(iface, flags);
 }
 
 static unsigned uct_rc_mlx5_iface_progress_cyclic(void *arg)
 {
-    return uct_rc_mlx5_iface_progress(arg, UCT_RC_MLX5_POLL_FLAG_HAS_EP);
+    return uct_rc_mlx5_iface_progress(arg, UCT_IB_MLX5_POLL_FLAG_HAS_EP);
+}
+
+static unsigned uct_rc_mlx5_iface_progress_zip(void *arg)
```
Comment (Artemy-Mellanox) [inline src/uct/ib/rc/accel/rc_mlx5_iface.c:187]: uct_rc_mlx5_iface_progress_cyclic_zip

File: src/uct/ib/rc/accel/rc_mlx5_iface.c [hunk 3]
```diff
@@ -606,7 +615,11 @@ uct_rc_mlx5_iface_init_rx(uct_rc_iface_t *rc_iface,
     if (iface->config.srq_topo == UCT_RC_MLX5_SRQ_TOPO_LIST) {
         iface->super.progress = uct_rc_mlx5_iface_progress_ll;
     } else {
-        iface->super.progress = uct_rc_mlx5_iface_progress_cyclic;
+        if (iface->cq[UCT_IB_DIR_RX].zip || iface->cq[UCT_IB_DIR_TX].zip) {
```
Comment (Artemy-Mellanox) [inline src/uct/ib/rc/accel/rc_mlx5_iface.c:618]: maybe avoid nesting

File: src/uct/ib/rc/base/rc_iface.c [hunk 4]
```diff
@@ -574,7 +574,7 @@ UCS_CLASS_INIT_FUNC(uct_rc_iface_t, uct_iface_ops_t *tl_ops,
 
     tx_cq_size                  = uct_ib_cq_size(&self->super, init_attr,
                                                  UCT_IB_DIR_TX);
-    self->tx.cq_available       = tx_cq_size - 1;
+    self->tx.cq_available       = tx_cq_size - 2;
```
Comment (Artemy-Mellanox) [inline src/uct/ib/rc/base/rc_iface.c:577]: add comment why

File: test/gtest/uct/ib/test_cqe_zipping.cc [hunk 5]
```diff
@@ -41,7 +41,7 @@ class test_cqe_zipping : public test_uct_ib_with_specific_port {
          * Local deadline prevents hanging inside receive loop in case of
          * connectivity issues
          */
-        const ucs_time_t deadline = ucs::get_deadline(60);
+        const ucs_time_t deadline = ucs::get_deadline(600);
```
Comment (Artemy-Mellanox) [inline test/gtest/uct/ib/test_cqe_zipping.cc:44]: comment that need longer time to wait for zip to happen

File: test/gtest/uct/ib/test_cqe_zipping.cc [hunk 6]
```diff
@@ -83,6 +86,17 @@ class test_cqe_zipping : public test_uct_ib_with_specific_port {
         m_send_cnt = m_recv_cnt = 0;
     }
 
+    bool check_ib_mlx5_md_caps(uint64_t flags) {
```
Comment (Artemy-Mellanox) [inline test/gtest/uct/ib/test_cqe_zipping.cc:89]: do not pass flags param
rename func to smth like check_cqe_zip_caps

File: src/uct/ib/mlx5/ib_mlx5.inl [hunk 7]
```diff
@@ -96,40 +108,50 @@ uct_ib_mlx5_check_and_init_zipped(uct_ib_mlx5_cq_t *cq, struct mlx5_cqe64 *cqe)
...
-static UCS_F_ALWAYS_INLINE struct mlx5_cqe64*
-uct_ib_mlx5_poll_cq(uct_ib_iface_t *iface, uct_ib_mlx5_cq_t *cq, int poll_flags)
+typedef struct mlx5_cqe64 *
+(uct_ib_mlx5_check_compl_cb_t)(uct_ib_iface_t *iface, uct_ib_mlx5_cq_t *cq,
+                               struct mlx5_cqe64 *cqe, int poll_flags);
+
+
+static UCS_F_ALWAYS_INLINE struct mlx5_cqe64 *
+uct_ib_mlx5_poll_cq_common(uct_ib_iface_t *iface, uct_ib_mlx5_cq_t *cq,
+                           int poll_flags, uct_ib_mlx5_check_compl_cb_t cb)
```
Comment (yosefe) [inline src/uct/ib/mlx5/ib_mlx5.inl:123]: rename cb -> check_cqe_cb

File: src/uct/ib/mlx5/ib_mlx5.inl [hunk 8]
```diff
@@ -96,40 +108,50 @@ uct_ib_mlx5_check_and_init_zipped(uct_ib_mlx5_cq_t *cq, struct mlx5_cqe64 *cqe)
...
 
     ucs_memory_cpu_load_fence();
 
     if (ucs_unlikely(uct_ib_mlx5_cqe_is_error_or_zipped(cqe->op_own))) {
-        return uct_ib_mlx5_check_completion(iface, cq, cqe);
+        return cb(iface, cq, cqe, poll_flags);
+    }
+
+    if (poll_flags & UCT_IB_MLX5_POLL_FLAG_CQE_ZIP) {
+        cq->cq_unzip.title_cqe_valid = 0;
```
Comment (yosefe) [inline src/uct/ib/mlx5/ib_mlx5.inl:142]: pls add a comment why

File: src/uct/ib/mlx5/ib_mlx5.inl [hunk 9]
```diff
@@ -96,40 +108,50 @@ uct_ib_mlx5_check_and_init_zipped(uct_ib_mlx5_cq_t *cq, struct mlx5_cqe64 *cqe)
...
-    cq->cq_unzip.zipped_block_seq_flag = 0;
-    cq->cq_ci = cqe_index + 1;
-    return cqe;
+    cq->cq_ci = idx + 1;
+    return cqe; /* TODO optimize - let complier know cqe is not null */
+}
+
+
+static UCS_F_ALWAYS_INLINE struct mlx5_cqe64*
+uct_ib_mlx5_poll_cq(uct_ib_iface_t *iface, uct_ib_mlx5_cq_t *cq, int poll_flags)
```
Comment (yosefe) [inline src/uct/ib/mlx5/ib_mlx5.inl:151]: maybe can remove this wrapper

File: test/gtest/uct/ib/test_cqe_zipping.cc (hunk 5 above)
Comment (yosefe) [reply test/gtest/uct/ib/test_cqe_zipping.cc:44]: unhandled

File: src/uct/ib/mlx5/ib_mlx5.inl [hunk 10]
```diff
@@ -18,9 +18,8 @@ typedef struct uct_ib_mlx5_wqe_ctrl_seg {
 static UCS_F_ALWAYS_INLINE UCS_F_NON_NULL struct mlx5_cqe64*
 uct_ib_mlx5_get_cqe(uct_ib_mlx5_cq_t *cq,  unsigned cqe_index)
 {
-    return UCS_PTR_BYTE_OFFSET(cq->cq_buf,
-                               ((cqe_index & UCS_MASK(cq->cq_length_log)) <<
-                                cq->cqe_size_log));
+    return UCS_PTR_BYTE_OFFSET(cq->cq_buf, (cqe_index & cq->cq_length_mask) <<
```
Comment (yosefe) [inline src/uct/ib/mlx5/ib_mlx5.inl:21]: newline after ,

**Review by yosefe — APPROVED** [review]

**Pull request merged**